### PR TITLE
Some fixes and performance stuff

### PR DESCRIPTION
**Pull request by ThatDan123**
Some performance stuff and NRE fixes

### CODE REVIEW

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs [hunk 1]
```diff
@@ -76,7 +78,12 @@ private float DealDamageAt(float damage, AttackType attackType, Vector3Int cellP
 		}
 
 		data.AddTileDamage(Layer.LayerType, basicTile.Armor.GetDamage(damage < basicTile.damageDeflection? 0: damage, attackType));
-		SoundManager.PlayNetworkedAtPos(basicTile.SoundOnHit, worldPosition);
+
+		if (basicTile.SoundOnHit.IsReadyLoaded)
+		{
+			SoundManager.PlayNetworkedAtPos(basicTile.SoundOnHit, worldPosition);
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs]: Bod, this probably isnt the best way, as im not sure if this blocks the sound from trying to load the first time, so would never to true. The issue was that it was spamming load but wasnt loading for some reason, it wasnt null or anything so is a bit weird.

**Comment by corp-0** [comment]
test is failing

File: UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs [hunk 2]
```diff
@@ -34,18 +34,7 @@ public partial class MatrixManager : MonoBehaviour
...
-		{
-			if (matrixManager == null)
-			{
-				matrixManager = FindObjectOfType<MatrixManager>();
-			}
-
-			return matrixManager;
-		}
-	}
+	public static MatrixManager Instance => matrixManager;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:37]: I think FindObjectOfType is required for the singleton monobehaviours

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:37]: nope

File: UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs [hunk 3]
```diff
@@ -79,6 +68,18 @@ private IEnumerator WaitForLoad()
...
 
+	private void Awake()
+	{
+		if (matrixManager == null)
+		{
+			matrixManager = this;
+		}
+		else
+		{
+			Destroy(this);
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:81]: could this be the thing making tests go ape?

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:81]: yup, cus of execute in edit mode, trying to find a way to fix

File: UnityProject/Assets/Scripts/Objects/ExosuitFabricator.cs [hunk 4]
```diff
@@ -62,10 +57,17 @@ private void Awake()
...
 		public void EnsureInit()
 		{
-			SyncSprite(ExosuitFabricatorState.Idle, ExosuitFabricatorState.Idle);
+			spriteHandler = GetComponentInChildren<SpriteHandler>();
 			materialStorage = this.GetComponent<MaterialStorage>();
 		}
 
+		private void Start()
+		{
+			if(CustomNetworkManager.IsServer == false) return;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/ExosuitFabricator.cs]: I remember there was a Server version of this event method, so you get rid of the if.

File: UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs [hunk 5]
```diff
@@ -64,16 +64,18 @@ public bool HasObjects(Vector3Int position)
...
 	{
-		return _objects.ContainsKey(position) ? _objects[position] : emptyList;
+		return _objects.TryGetValue(position, out var objectsOut) ? objectsOut : emptyList;
 	}
 
-	public IEnumerable<RegisterTile> Get(Vector3Int position, ObjectType type) {
-		if ( !HasObjects( position ) )
+	public IEnumerable<RegisterTile> Get(Vector3Int position, ObjectType type)
+	{
+		if (_objects.TryGetValue(position, out var objectsOut) == false)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs:72]:
```suggestion
		if (_objects.TryGetValue(position, out _) == false)
```
If you're not using the out value, you can dispose it like this.

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs:72]: It is using the out value

File: UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs (hunk 3 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:81]: Added a check which fixes this

File: UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs (hunk 5 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs:72]: ah yeah, got confused by the return brackets.

File: UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs (hunk 2 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:37]: Most managers should be moved over to this system, as checking FindObjectOfType is expensive for start up. But the main reason for me switching it up was the null check on the get is very bad.

Comment (corp-0) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:37]: Let's wait on what @Bod9001 has to say before marking as resolved.

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs (hunk 1 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs]: you could use a try catch to see if it's doing anything?

File: UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs:37]: This should work okay as long as it's not used before Awake, you can use the initialisation manager to ensure that it is so

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs (hunk 1 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/TilemapDamage.cs]: Ive removed the check for now

**Comment by Aranclanos** [comment]
are you sure about some of the checks? I didnt look deeply but it feels like the pipe and disposal ones are hiding a more in depth problem

**Comment by ThatDan123** [comment]
The disposal one is just checking to see if its null when the object is destroyed, dont see why that one could be that bad to add.

The pipe check i know less about the system, but it seems like thats right as it should return its internal mixture if its not on the network?

**Comment by corp-0** [comment]
are we ready to merge?

**Comment by ThatDan123** [comment]
yeah, i think

**Comment by corp-0** [comment]
boom, merged

**Pull request merged**